### PR TITLE
solvers: modify symbols ordering in _solve_system

### LINKED ISSUE #18208: linsolve gets confused with structured block matrix (underdetermined, redundant rows)
Hello, I tried to use linsolve for solving a puzzle but I think I ran into a bug.  I could get a proper solution with the solve function though.

So I have two sets of variables x* and y*, each related through redundant equations.  I also have a partial solution, crossing the bound between x* and y*, I put that into partsol.  Putting it all together gives different results for linsolve and solve.
Classic solve gives the right answer, but linsolve seems to mix x-variables into the y-equations.  I see y7 where x3 (==x1+10) should be, x3 for y9 and y9 for y7 (and maybe more).
Maybe this is related to #14594?

P.S.: You may have heard this before, but sympy is cool software.  Thanks for it!

<details><summary>code and results</summary><p>

```
#! /usr/bin/python3

from sympy import symbols, solve, linsolve, linear_eq_to_matrix

vars = symbols('x0:16') + symbols('y0:12')
x0, x1, x2, x3, x4, x5, x6, x7, x8, x9, x10, x11, x12, x13, x14, x15,\
  y0, y1, y2, y3, y4, y5, y6, y7, y8, y9, y10, y11 = vars

eqs = [x0 + x1 + x2 + x3 - 51,
 x0 + x1 + x4 + x5 - 46,
 x2 + x3 + x6 + x7 - 39,
 x0 + x3 + x4 + x7 - 50,
 x1 + x2 + x5 + x6 - 35,
 x4 + x5 + x6 + x7 - 34,
 x4 + x5 + x8 + x9 - 46,
 x10 + x11 + x6 + x7 - 23,
 x11 + x4 + x7 + x8 - 25,
 x10 + x5 + x6 + x9 - 44,
 x10 + x11 + x8 + x9 - 35,
 x12 + x13 + x8 + x9 - 35,
 x10 + x11 + x14 + x15 - 29,
 x11 + x12 + x15 + x8 - 35,
 x10 + x13 + x14 + x9 - 29,
 x12 + x13 + x14 + x15 - 29,
 y0 + y1 + y2 + y3 - 55,
 y0 + y1 + y4 + y5 - 53,
 y2 + y3 + y6 + y7 - 56,
 y0 + y3 + y4 + y7 - 57,
 y1 + y2 + y5 + y6 - 52,
 y4 + y5 + y6 + y7 - 54,
 y4 + y5 + y8 + y9 - 48,
 y10 + y11 + y6 + y7 - 60,
 y11 + y4 + y7 + y8 - 51,
 y10 + y5 + y6 + y9 - 57,
 y10 + y11 + y8 + y9 - 54]

partsol = [x10 - 2, x11 - 5, x12 - 1, x13 - 6, x14 - 1, x15 - 21, y0 - 12, y1 - 20]

print('matrix:')
print(linear_eq_to_matrix(eqs+partsol, vars))

print('\nlinsolve gives:')
linsolution = linsolve(eqs+partsol, vars)
for (var, sol) in zip(vars, next(iter(linsolution))):
  print(f"{var} = {sol}")

print('\nsolve gives:')
solution = solve(eqs+partsol, vars)
for v in vars:
  if v in solution:
    print(f"{v} = {solution[v]}")
  else:
    print(f"{v} = {v}")

#may be handy:
asolution = {x0: 20, x1: 8, x2: 5, x3: 18, x4: 5, x5: 13, x6: 9, x7: 7,
             x8: 8, x9: 20, x10: 2, x11: 5, x12: 1, x13: 6, x14: 1,
             x15: 21, y0: 12, y1: 20, y2: 9, y3: 14, y4: 12,
             y5: 9, y6: 14, y7: 19, y8: 15, y9: 12, y10: 22, y11: 5}

** OUPUT **
matrix:
(Matrix([
[1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), Matrix([
[51],
[46],
[39],
[50],
[35],
[34],
[46],
[23],
[25],
[44],
[35],
[35],
[29],
[35],
[29],
[29],
[55],
[53],
[56],
[57],
[52],
[54],
[48],
[60],
[51],
[57],
[54],
[ 2],
[ 5],
[ 1],
[ 6],
[ 1],
[21],
[12],
[20]]))

linsolve gives:
x0 = 38 - y7
x1 = y7 - 10
x2 = 23 - y7
x3 = x3
x4 = 12 - x7
x5 = x7 + 6
x6 = 16 - x7
x7 = x7
x8 = 8
x9 = 20
x10 = 2
x11 = 5
x12 = 1
x13 = 6
x14 = 1
x15 = 21
y0 = 12
y1 = 20
y2 = x3 - y11 + 2
y3 = -x3 + y11 + 21
y4 = x3 - y11 - y9 + 24
y5 = -x3 + y11 + y9 - 3
y6 = 33 - y9
y7 = y7
y8 = 27 - x3
y9 = y9
y10 = 27 - y11
y11 = y11

solve gives:
x0 = 28 - x1
x1 = x1
x2 = 13 - x1
x3 = x1 + 10
x4 = 18 - x5
x5 = x5
x6 = 22 - x5
x7 = x5 - 6
x8 = 8
x9 = 20
x10 = 2
x11 = 5
x12 = 1
x13 = 6
x14 = 1
x15 = 21
y0 = 12
y1 = 20
y2 = -y11 + y9 + 2
y3 = y11 - y9 + 21
y4 = -y11 - y7 + y9 + 24
y5 = y11 + y7 - y9 - 3
y6 = 33 - y7
y7 = y7
y8 = 27 - y9
y9 = y9
y10 = 27 - y11
y11 = y11
```

</p></details>

### PR DESCRIPTION
**Pull request by AaronStiff**
#### References to other Issues or PRs

Fixes #18208 

#### Brief description of what is fixed or changed
`solve` was solving a system of equations in a different, albeit valid, way to `linsolve`. Modified the way `solvers._solve_system` sorts symbols to preserve order and keep `solve` and `linsolve` in sync.

#### Other comments
Will be adding tests for other solvers to test the same system of equations which caused the issue.

#### Release Notes

* solvers
  * Modified symbols sorting in solvers._solve_system to ensure solve and linsolve solve the same way

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
The release note should be for solvers rather than polys.

I think it would be better to put these tests together in one place so that the code for the system and solution is not duplicated and to emphasise the fact that we are asserting that all solvers give the same results.

**Comment by AaronStiff** [comment]
Oh, whoops. Sure thing. 

Just put them all in `test_solvers` then?

**Comment by oscarbenjamin** [comment]
> Just put them all in `test_solvers` then?

I would put them wherever the tests for linsolve are.

**Comment by AaronStiff** [comment]
In `test_solveset`. Okay

File: sympy/solvers/tests/test_solveset.py [hunk 1]
```diff
@@ -2490,6 +2491,54 @@ def test_issue_10426():
...
+    # gauss_jordan_solve
+    lhs, rhs = linear_eq_to_matrix(eqs+partsol, vars)
+    gj_solve, new_vars = lhs.gauss_jordan_solve(rhs)
+    gj_solve = [i for i in gj_solve]
+    tau0, tau1, tau2, tau3, tau4 = symbols([str(v) for v in new_vars])
+
+    gj_expected = [38 - tau0, tau0 - 10, 23 - tau0, tau0, 12 - tau1, tau1 + 6, 16 - tau1,\
+                tau1, 8, 20, 2, 5, 1, 6, 1, 21, 12, 20, tau3 - tau4 + 2, -tau3 + tau4 + 21,\
+                -tau2 + tau3 - tau4 + 24, tau2 - tau3 + tau4 - 3, 33 - tau2, tau2,\
+                27 - tau3, tau3, 27 - tau4, tau4]
```
Comment (oscarbenjamin) [inline sympy/solvers/tests/test_solveset.py:2539]: Can we make this expected solution from the other one using subs to replace the symbols?

The idea is to be clear that it is the same solution but just with different symbol names.

Comment (oscarbenjamin) [reply sympy/solvers/tests/test_solveset.py:2539]: Like `gj_expected = linsolve_expected.subs(zip([x3, ...], vars))`.

Comment (AaronStiff) [reply sympy/solvers/tests/test_solveset.py]: Sure. Is there a cleaner way than this?

```py
gj_expected = list(list(linsolve_expected.subs(zip([x3, x7, y7, y9, y11], [tau0, tau1, tau2, tau3, tau4])))[0])
```

Comment (oscarbenjamin) [reply sympy/solvers/tests/test_solveset.py]: Does the code I showed not work?

Comment (AaronStiff) [reply sympy/solvers/tests/test_solveset.py]: Oh sorry, for some reason I didn't see that when I replied.

No, it doesn't seem to work:

```pycon
        gj_expected = linsolve_expected.subs(zip([x3, x7, y7, y9, y11], new_vars))
>       assert gj_solve == gj_expected
E       assert [38 - tau0, t...tau1 + 6, ...] == FiniteSet((38...- tau4, tau4))
E         Use -v to get the full diff

test_solveset.py:2538: AssertionError
```

`linsolve_expected` is a tuple in a FiniteSet, whereas `gj_solve` is just a list.

Comment (oscarbenjamin) [reply sympy/solvers/tests/test_solveset.py]: Oh, sorry I see what you mean. Yes, I guess what you have is fine. You can also convert the other way like
```
assert FiniteSet(Tuple(gj_solve)) == gj_expected
```

Comment (oscarbenjamin) [reply sympy/solvers/tests/test_solveset.py]: I would break up the long line though. Perhaps do the subs in one line and the other parts in another.

Comment (AaronStiff) [reply sympy/solvers/tests/test_solveset.py]: Actually converting `gj_solve` looks quite nice. I'll do that.

**Comment by AaronStiff** [comment]
Why is `test_elliptic_integrals` failing? When I test it on my own machine, it passes. I noticed on some of the past tests that `vector.frame.py` was also failing, but not this time round.

**Comment by oscarbenjamin** [comment]
The test is a random test so it can potentially fail for some inputs but not necessarily in a reproducible way.

**Comment by AaronStiff** [comment]
I see. How are failed random tests like this one usually treated? Will the issue have to be debugged somehow?

**Comment by oscarbenjamin** [comment]
It's probably not possible to reproduce this. At the top of the test output you can see the random seed as well as the PYTHONHASHSEED which controls set ordering. In principle you can set those seeds when running the tests like:
```
PYTHONHASHSEED=454480227 bin/test sympy/functions/special/tests/test_elliptic_integrals.py -k test_P --seed=2487161
```
However the random seed is not set at the start of each separate test so to actually reproduce this you would need to run the exact same tests that `bin/test` did like:
```
PYTHONHASHSEED=454480227 bin/test --seed=2487161 --split=1/2
```
That runs half of the test suite and the failing test won't come up for some time. In between other tests may or may not be skipped due to the presence or absence of different libraries or Python versions etc which might alter the random seed by the time you get to the one that failed.

Without knowing what the seed was I just tried running this random test many times:
```python
from sympy.testing.randtest import (test_derivative_numerically as td,
                                      random_complex_number as randcplx)
from sympy.functions.special.elliptic_integrals import elliptic_pi as P
from sympy import srepr

from sympy.abc import m

for i in range(10000):
    print(i)
    rx, ry = randcplx(), randcplx()
    assert td(P(rx, ry, m), m), str((srepr(rx), srepr(ry)))
```
I think I let it go about 5000 times and didn't see it fail once. That's why we really need to know the seed when something like this happens.

There are two approaches that would help with this sort of thing in future (ideally we should do both of these):

1. A randomised test like this should print out a complete representation of any random values on failure like:

        x = Float(random())
        assert f(x) < x, srepr(x)

2. The test runner should reset the random seed before each individual test so that rerunning an individual test with the same `--seed` value is more likely to reproduce the random failure.

If you feel like fixing either of those then you can push up a commit to do that. Otherwise you can close and reopen the PR which will rerun the tests and I expect that they will pass.

**Comment by AaronStiff** [comment]
I think I'll just close and reopen. I'm still working out why `nonlinsolve` is returning a slightly different solution to `linsolve`.

**Comment by oscarbenjamin** [comment]
Looks like the same test failed again. Let's see what happens after another commit is pushed...

**Comment by AaronStiff** [comment]
Sure. I just added a blank line.

**Comment by AaronStiff** [comment]
@oscarbenjamin, `nonlinsolve` is returning slightly different results to `linsolve`, but not in the same way `solve` was:

```
linsolve:
FiniteSet((38 - x3, x3 - 10, 23 - x3, x3, 12 - x7, x7 + 6, 16 - x7, x7, 8, 20, 2, 5, 1, 6, 1, 21, 12, 20, -y11 + y9 + 2, y11 - y9 + 21, -y11 - y7 + y9 + 24, y11 + y7 - y9 - 3, 33 - y7, y7, 27 - y9, y9, 27 - y11, y11))

nonlinsolve:
FiniteSet((38 - x3, x3 - 10, 23 - x3, x3, 12 - x7, x7 + 6, 16 - x7, x7, 8, 20, 2, 5, 1, 6, 1, 21, 12, 20, -y5 + y7 - 1, y5 - y7 + 24, 21 - y5, y5, 33 - y7, y7, 27 - y9, y9, -y5 + y7 - y9 + 24, y5 - y7 + y9 + 3))
```

It seems only equations with `y` terms are changed.

The problem is that I don't think `nonlinsolve` uses the same process as any of the other 3 solvers I've added tests for. It doesn't solve or choose free symbols using RREF. Is it still something to try to harmonize, or should it be left out for now? I'm kind of at a dead end at the moment.

**Comment by oscarbenjamin** [comment]
> The problem is that I don't think `nonlinsolve` uses the same process as any of the other 3 solvers I've added tests for. It doesn't solve or choose free symbols using RREF.

No, I see that now. I think that probably the `substitution` function when applied to a linear system is more or less equivalent to Gaussian elimination but it doesn't necessarily do the operations in the same order so it can lead to different symbols being given here.

Probably nonlinsolve should pass any subsystems of linear equations to linsolve but that's a separate issue.

> Is it still something to try to harmonize, or should it be left out for now? I'm kind of at a dead end at the moment.

For now I would say add a test along with the others showing the different output that nonlinsolve gives but add code comments clearly explaining that ideally nonlinsolve would give the same outputs as the others. Then if the output from nonlinsolve changes later (e.g. because it is made to use linsolve) the test will fail but the comment will make it clear that it's okay to change the test.

**Comment by AaronStiff** [comment]
Makes sense!

**Comment by AaronStiff** [comment]
> Actually above I would use brackets like:
> 
> ```
>     (x0, x1, x2, x3, x4, x5, x6, x7, x8, x9, x10, x11, x12, x13, x14, x15,
>     y0, y1, y2, y3, y4, y5, y6, y7, y8, y9, y10, y11) = vars
> ```
>
> Then again:
> 
> ```
>     x0, x1, x2, x3, x4, x5, x6, x7, x8, x9, x10, x11, x12, x13, x14, x15 == symbols('x:16')
>     y0, y1, y2, y3, y4, y5, y6, y7, y8, y9, y10, y11 = symbols('y:12')
> ```

In that order? I get a TypeError saying it can't unpack  a non-iterable.

> The equations would be easier to read if you split a new line after each comma like:
> 
> ```
>    eqs = [
>         x0 + x1 + x2 + x3 - 51,
>         x0 + x1 + x4 + x5 - 46,
>         x2 + x3 + x6 + x7 - 39,
>         ...
> ```

Sure thing.

**Comment by AaronStiff** [comment]
> There's no need for partsol to be a separate list

Got it.

**Comment by AaronStiff** [comment]
> `linear_eq_to_matrix` was already called above. I suggest to do it once like
> 
> ```
> A, b = linear_eq_to_matrix(eqs+partsol, vars)
> assert linsolve(matrix, vars) == linsolve_expected
> assert A.gauss_jordan ...
> ```

Yes, that simplifies it.

**Comment by AaronStiff** [comment]
> This message should be clearer: the solution set is equivalent and is also correct. However we would prefer to use the same symbols as parameters for the solution to the underdetermined system in all cases if possible. So what we want is a solution that is not just equivalent but also given in the same form.

Sure. I kind of just copied what you wrote and modified it a bit. :)

**Comment by AaronStiff** [comment]
> This is the same as `linsolve_expected`. Can we derive one from the other?

Which is the same? `nonlinsolve_expected` is different.

**Comment by oscarbenjamin** [comment]
> > This is the same as `linsolve_expected`. Can we derive one from the other?
> 
> Which is the same? `nonlinsolve_expected` is different.

It looks like I accidentally commented on the commit rather than the diff in the PR. I think I was referring to `solve_expected`.

**Comment by oscarbenjamin** [comment]
When squashing commits in a PR it's best just to do it at the end (if needed at all). While working through it gets confusing if the commits commented on disappear and it's also helpful to be able to see any test failures etc - those disappear once you force push.

**Comment by AaronStiff** [comment]
Oh, okay. Sorry about that.

I think it could probably be derived, yes.

**Comment by AaronStiff** [comment]
That force push was just some spaces I forgot to add. :)

**Comment by oscarbenjamin** [comment]
Okay, looks good to me. I can't merge this because it's marked as a "draft" though.

**Comment by AaronStiff** [comment]
Great. I'll update to a regular PR.

**Comment by oscarbenjamin** [comment]
The tests failed again due to #20933

**Comment by oscarbenjamin** [comment]
Looks good

**Pull request merged**